src/api-service/__app__/agent_events/__init__.py

```diff
@@ -82,6 +88,17 @@ def on_state_update(
                             state=NodeTaskState.setting_up,
                         )
                         node_task.save()
+            elif state == NodeState.done:
+                # Model-validated.
+                done_data = cast(NodeDoneEventData, state_update.data)
+
+                if done_data.error:
+                    logging.error(
+                        "node done with error disposition: machine_id = %s, data = %s",
+                        machine_id,
+                        done_data,
+                    )
+                    # TODO: update task states linked to node.
```

Comment:
Other than logging do we want to do something here, or let the normal `done` machinery pick up the failure?

Reply:
If we want to indicate task + node-task failure immediately, this is where we'd do it. Because we've established the node-task link in the `setting_up` event, we can query table storage for the tasks that were assigned to the node.

Question: do we want to also make sure we stop _other_ occurrences of the task on other nodes (if `vm_count` > 1), or even the job? Probably, because we've discussed "for tasks with multiple `vm_count`, the state should be the most-progressed or most-failed state".